### PR TITLE
checking /proc/sys/kernel/ftrace_enabled

### PR DESCRIPTION
**Pull request by mtcherni95**
Resolves: #1150

### CODE REVIEW

**Comment by itaysk** [comment]
> Resolves: #1150

this goes in the PR body, not title :)

**Comment by itaysk** [comment]
In the linked issue you say you will add the check to libbpfgo, what changed?

**Comment by rafaeldtinoco** [comment]
Yes, @mtcherni95 , please check:

https://github.com/aquasecurity/libbpfgo/pull/94

**Comment by mtcherni95** [comment]
I will update PR by using helper in libbpfgo - this will require update of go mod version of libbpgo

**Comment by rafaeldtinoco** [comment]
> I will update PR by using helper in libbpfgo - this will require update of go mod version of libbpgo

Would you include:

```diff
commit eb264af (HEAD -> main)
Author: Rafael David Tinoco <rafaeldtinoco@gmail.com>
Date:   Fri Nov 5 20:01:05 2021

    tracee: new AttachTracepoint prototype

diff --git a/tracee-ebpf/tracee/tracee.go b/tracee-ebpf/tracee/tracee.go
index ee1d08e..72d3992 100644
--- a/tracee-ebpf/tracee/tracee.go
+++ b/tracee-ebpf/tracee/tracee.go
@@ -868,7 +868,8 @@ func (t *Tracee) initBPF() error {
                        case kretprobe:
                                _, err = prog.AttachKretprobe(probe.event)
                        case tracepoint:
-                               _, err = prog.AttachTracepoint(probe.event)
+                               tpEvent := strings.Split(probe.event, ":")
+                               _, err = prog.AttachTracepoint(tpEvent[0], tpEvent[1])
                        case rawTracepoint:
                                tpEvent := strings.Split(probe.event, ":")[1]
                                _, err = prog.AttachRawTracepoint(tpEvent)
```

to tracee as well ? You will need when bumping libbpfgo version.

**Comment by grantseltzer** [comment]
@mtcherni95 Now that we've upgraded the libbpfgo dependency to the latest release (which includes this helper function) you can update this PR.

**Comment by mtcherni95** [comment]
> @mtcherni95 Now that we've upgraded the libbpfgo dependency to the latest release (which includes this helper function) you can update this PR.

great, thanks a lot

File: tracee-ebpf/main.go [hunk 1]
```diff
@@ -95,13 +95,19 @@ func main() {
...
+			if err = checkRequiredCapabilities(selfCap); err != nil {
+				return err
+			}
+
+			enabled, err := helpers.FtraceEnabled()
 			if err != nil {
 				return err
 			}
+			if !enabled {
+				fmt.Fprintf(os.Stderr, "ftrace_enabled: warning: ftrace is not enabled - kprobes won't be inserted\n")
```
Comment (grantseltzer) [inline tracee-ebpf/main.go:107]: To a user, I don't know if having "kprobes won't be inserted" is very useful. Is it possible to list events that won't be traced? Are we sure that ftrace is absolutely required for all kprobes?

Comment (mtcherni95) [reply tracee-ebpf/main.go:107]: I think so yeah. @yanivagman ?

Comment (rafaeldtinoco) [reply tracee-ebpf/main.go:107]: Ftrace is needed for kprobes as they're built on top of ftrace base architecture (https://www.kernel.org/doc/Documentation/trace/ftrace-design.txt), like an extension to it. I don't think we should expand to much here, but I do think the message could be clearer for an average user that might not know what kprobes and ftrace are.

Comment (mtcherni95) [reply tracee-ebpf/main.go:107]: I actually think that if a user is using Tracee then most likely he will have some technical background (and kprobes knowledge as well). I suggest something like `"ftrace_enabled: warning:  ftrace is not enabled - some events may not be caught (kprobes not enabled)"`.

Comment (rafaeldtinoco) [reply tracee-ebpf/main.go:107]: I believe no events, at all, will be caught, if ftrace is disabled. 

And I disagree that users of tracee will know what a kprobe (or ftrace) is. 

I do agree that a warning like that could be helpful as long as it is rephrased to something like:

`ftrace_enabled: warning: ftrace is not enabled, kernel events won't be caught, make sure to enable it by executing echo 1 | sudo tee /proc/sys/kernel/ftrace_enabled`

Comment (mtcherni95) [reply tracee-ebpf/main.go:107]: You do get events coming from tracepoints. 

Let's go with your message then :)

Comment (rafaeldtinoco) [reply tracee-ebpf/main.go]: > You do get events coming from tracepoints.

I see what you mean. But our ebpf logic might be broken as there are maps manipulated by different eBPF programs (with different probe methods) in different times, I believe. 

I agree that if we consider tracee purely as an introspection tool, final user might know what we're talking about... but we're trying to shift that mindset as well (thus the explicit message about how to get it enabled).

Cool, thanks.

Comment (grantseltzer) [reply tracee-ebpf/main.go]: So if this is the case should we just error out if ftrace is disabled? Is there any point in running tracee?

Comment (yanivagman) [reply tracee-ebpf/main.go]: > So if this is the case should we just error out if ftrace is disabled? Is there any point in running tracee?

I don't think so - tracepoint will still work, which means that all the syscalls + some other events will still work (> 95% of the events are not kprobes) - so I think we should just warn

Comment (rafaeldtinoco) [reply tracee-ebpf/main.go]: I believe you're right @grantseltzer. Its better to just fail instead of trying to rely in a tainted execution.

Comment (rafaeldtinoco) [reply tracee-ebpf/main.go]: Alright then, works for me.

**Review by grantseltzer — APPROVED** [review]
LGTM

**Review by yanivagman — APPROVED** [review]

**Pull request merged**